doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -105,197 +129,222 @@ simply :g:`t=u` dropping the implicit type of :g:`t` and :g:`u`.
       the same key as the left- or right-hand side of the lemma given to rewrite, and the arguments
       are then unified up to full reduction.
 
-.. tacn:: replace @term with @term’
-   :name: replace
-
-   This tactic applies to any goal. It replaces all free occurrences of :n:`@term`
-   in the current goal with :n:`@term’` and generates an equality :n:`@term = @term’`
-   as a subgoal. This equality is automatically solved if it occurs among
-   the assumptions, or if its symmetric form occurs. It is equivalent to
-   :n:`cut @term = @term’; [intro H`:sub:`n` :n:`; rewrite <- H`:sub:`n` :n:`; clear H`:sub:`n`:n:`|| assumption || symmetry; try assumption]`.
-
-   .. exn:: Terms do not have convertible types.
-      :undocumented:
-
-   .. tacv:: replace @term with @term’ by @tactic
-
-      This acts as :n:`replace @term with @term’` but applies :token:`tactic` to solve the generated
-      subgoal :n:`@term = @term’`.
+.. tacn:: rewrite * {? {| -> | <- } } @one_term {? in @ident } {? at @rewrite_occs } {? by @ltac_expr3 }
+          rewrite * {? {| -> | <- } } @one_term at @rewrite_occs in @ident {? by @ltac_expr3 }
+   :name: rewrite *; _
+   :undocumented:
 
-   .. tacv:: replace @term
+.. tacn:: rewrite_db @ident {? in @ident }
+   :undocumented:
 
-      Replaces :n:`@term` with :n:`@term’` using the first assumption whose type has
-      the form :n:`@term = @term’` or :n:`@term’ = @term`.
+.. tacn:: replace @one_term__from with @one_term__to {? @occurrences } {? by @ltac_expr3 }
+          replace {? {| -> | <- } } @one_term__from {? @occurrences }
+   :name: replace; _
 
-   .. tacv:: replace -> @term
+   The first form replaces all free occurrences of :n:`@one_term__from`
+   in the current goal with :n:`@one_term__to` and generates an equality
+   :n:`@one_term__to = @one_term__from`
+   as a subgoal. (Note the generated equality is reversed with respect
+   to the order of the two terms in the tactic syntax; see
+   issue `#13480 <https://github.com/coq/coq/issues/13480>`_.)
+   This equality is automatically solved if it occurs among
+   the hypotheses, or if its symmetric form occurs.
 
-      Replaces :n:`@term` with :n:`@term’` using the first assumption whose type has
-      the form :n:`@term = @term’`
+   The second form, with `->` or no arrow, replaces :n:`@one_term__from`
+   with :n:`@term__to` using
+   the first hypothesis whose type has the form :n:`@one_term__from = @term__to`.
+   If `<-` is given, the tactic uses the first hypothesis with the reverse form,
+   i.e. :n:`@term__to = @one_term__from`.
 
-   .. tacv:: replace <- @term
+   :n:`@occurrences`
+     The `type of` and `value of` forms are not supported.
```

Comment:
I just discovered (by testing) that in the case of `replace` too, you need to `Require Setoid` to be able to use an `at` clause.

Reply:
I added a sentence saying that, but the tactic description doesn't otherwise mention setoid equality.  So I guess the part "and generates an equality :n:`@one_term__to = @one_term__from` as a subgoal" needs to be rephrased similar to the wording for `rewrite`, i.e. with `EQ`?

Reply:
No, `replace` only works with Leibniz equality. The need to `Require Setoid` to be able to use an `at` clause seems to be a stupid limitation due to an implementation detail.

Reply:
What's the implementation detail limitation?

Reply:
The implementation detail is that `replace` uses `rewrite` internally and thus suffers from the same limitations (that the `at` clause is only implemented for `setoid_rewrite`).

Reply:
```coq
Require Setoid.
Set Typeclasses Debug.
Goal 0 = 0.
replace 0 with 1. (* no output *)
Undo.
replace 0 with 1 at 1. (* typeclass resolution output *)
```